tests/robustness/test_engine/engine.go

```diff
@@ -0,0 +1,192 @@
+	"github.com/kopia/kopia/tests/robustness/snapmeta"
+	"github.com/kopia/kopia/tests/tools/fio"
+	"github.com/kopia/kopia/tests/tools/fswalker"
+	"github.com/kopia/kopia/tests/tools/kopiarunner"
+)
+
+const (
+	// S3BucketNameEnvKey is the environment variable required to connect to a repo on S3
+	S3BucketNameEnvKey = "S3_BUCKET_NAME"
+)
+
+// ErrS3BucketNameEnvUnset is the error returned when the S3BucketNameEnvKey environment variable is not set
+var ErrS3BucketNameEnvUnset = fmt.Errorf("environment variable required: %v", S3BucketNameEnvKey)
+
+// Engine is the outer level testing framework for robustness testing
+type Engine struct {
+	FileWriter      *fio.Runner
+	TestRepo        snap.Snapshotter
+	MetaStore       snapmeta.Persister
+	Checker         *checker.Checker
+	cleanupRoutines []func()
+}
+
+// NewEngine instantiates a new Engine and returns its pointer. It is
+// currently created with:
+// - FIO file writer
+// - Kopia test repo snapshotter
+// - Kopia metadata storage repo
+// - FSWalker data integrity checker
+func NewEngine() (*Engine, error) {
+	e := new(Engine)
+
+	var err error
+
+	// Fill the file writer
+	e.FileWriter, err = fio.NewRunner()
+	if err != nil {
+		e.Cleanup() //nolint:errcheck
+		return nil, err
+	}
+
+	e.cleanupRoutines = append(e.cleanupRoutines, e.FileWriter.Cleanup)
+
+	// Fill Snapshotter interface
+	kopiaSnapper, err := kopiarunner.NewKopiaSnapshotter()
+	if err != nil {
+		e.Cleanup() //nolint:errcheck
+		return nil, err
+	}
+
+	e.cleanupRoutines = append(e.cleanupRoutines, kopiaSnapper.Cleanup)
+	e.TestRepo = kopiaSnapper
+
+	// Fill the snapshot store interface
+	snapStore, err := snapmeta.New()
+	if err != nil {
+		e.Cleanup() //nolint:errcheck
+		return nil, err
+	}
+
+	e.cleanupRoutines = append(e.cleanupRoutines, snapStore.Cleanup)
+
+	e.MetaStore = snapStore
+
+	// Create the data integrity checker
+	chk, err := checker.NewChecker(kopiaSnapper, snapStore, fswalker.NewWalkCompare())
+	e.cleanupRoutines = append(e.cleanupRoutines, chk.Cleanup)
+
+	if err != nil {
+		e.Cleanup() //nolint:errcheck
+		return nil, err
+	}
+
+	e.Checker = chk
+
+	return e, nil
+}
+
+// Cleanup cleans up after each component of the test engine
+func (e *Engine) Cleanup() error {
+	defer e.cleanup()
+
+	if e.MetaStore != nil {
+		return e.MetaStore.FlushMetadata()
+	}
+
+	return nil
+}
+
+func (e *Engine) cleanup() {
+	for _, f := range e.cleanupRoutines {
+		f()
+	}
+}
+
+// InitS3 attempts to connect to a test repo and metadata repo on S3. If connection
+// is successful, the engine is populated with the metadata associated with the
+// snapshot in that repo. A new repo will be created if one does not already
+// exist.
+func (e *Engine) InitS3(ctx context.Context, testRepoPath, metaRepoPath string) error {
+	bucketName := os.Getenv(S3BucketNameEnvKey)
+	if bucketName == "" {
+		return ErrS3BucketNameEnvUnset
+	}
+
+	err := e.MetaStore.ConnectOrCreateS3(bucketName, metaRepoPath)
```

Comment:
Does it have to be S3? Can this be made blob store agnostic?

IIRC given the appropriate configuration, kopia would create the bucket if it does not exist.

Reply:
Are you asking specifically about the metadata store repo? Or just connection to any repository?

The assumption and limitation of these two initialization helper functions (`eng.InitS3()` and `eng.InitFilesystem()`) is that the caller will connect/create a metadata repo that's in the same blob store provider as the test repo. For example if we want to connect to a filesystem test repo, we assume we'll find the metadata repo also in filesystem. Same for S3, where additionally the init function expects the metadata to live in the same bucket.

That said, these are all exported fields and methods, so if an Engine user wants to customize it where they have the metadata repo and test repo living in different buckets, or in different blob store services altogether, they can call `ConnectorCreate<blobstore>()` for each with the appropriate parameters.

These `ConnectOrCreate` functions are part of the `RepoManager` interface, which any repository (kopia, restic, whatever) must implement to be managed by the test engine (or at the very least return an error). Currently only the S3 and Filesystem funcs are added to the interface, but more can be added, to tell the engine to look for repos in google or wherever else.

Reply:
oh also

>IIRC given the appropriate configuration, kopia would create the bucket if it does not exist.

I don't think it does, unless I'm missing some required setup:

```
kopia: error: can't connect to storage: bucket "nick-kasten-io-test-2" does not exist, try --help
```